src/odemis/driver/actuator.py

```diff
@@ -940,6 +940,188 @@ def reference(self, axes):
+        except Exception as e:
+            self._child.stop({self._caxis})  # prevent any move queued
+            self.state._set_value(e, force_write=True)
+            logging.exception(e)
+
+    def _update_child_position(self, value):
+        pos = value[self._caxis] - self._offset
+        self.position._set_value({self._axis: pos}, force_write=True)
+
+    def _update_child_ref(self, value):
+        referenced = value[self._caxis]
+        self.referenced._set_value({self._axis: referenced}, force_write=True)
+
+    @isasync
+    def moveRel(self, shift):
+        """
+        Move the rotation actuator by a defined shift.
+        shift dict(string-> float): name of the axis and shift
+        """
+        if not shift:
+            return model.InstantaneousFuture()
+        self._checkMoveRel(shift)
+        f = self._executor.submit(self._doMoveRel, shift)
+
+        return f
+
+    def _doMoveRel(self, shift):
+        """
+        shift dict(string-> float): name of the axis and shift
+        """
+        cur_pos = self._child.position.value[self._caxis] - self._offset
+        pos = cur_pos + shift[self._axis]
+
+        # Reference from time to time
+        self._move_num += 1
+        if self._ref_period and self._move_num >= self._ref_period:
+            logging.debug("Referencing axis %s after %s moves." % self._caxis, self._ref_period)
+            self._child.reference({self._caxis}).result()
+            self._move_num = 0
+
+        self._doMoveAbs({self._axis: pos})
+
+    @isasync
+    def moveAbs(self, pos):
+        """
+        Move the actuator to the defined position in m for each axis given.
+        pos dict(string-> float): name of the axis and position in m
+        """
+
+        if not pos:
+            return model.InstantaneousFuture()
+        self._checkMoveAbs(pos)
+        pos = self._applyInversion(pos)
+        f = self._executor.submit(self._doMoveAbs, pos)
+
+        return f
+
+    def _doMoveAbs(self, pos):
+        axis, distance = pos.items()[0]
+
+        # Reference from time to time
+        self._move_num += 1
+        if self._ref_period and self._move_num >= self._ref_period:
+            logging.debug("Referencing axis %s (-> %s) after %s moves." % self._axis, self._caxis, self._ref_period)
+            self._child.reference({self._caxis}).result()
+            self._move_num = 0
+
+        logging.debug("Moving axis %s (-> %s) to %g", self._axis, self._caxis, distance)
+        move = {self._caxis: distance + self._offset}
+        self._child.moveAbs(move).result()
+
+    def _doReference(self, axes):
+        # Axis might always go towards negative direction during referencing. Make
+        # sure that the referencing works properly in case of a negative position (especially
+        # important if the actuator is cyclic).
+        logging.debug("Moving to reference starting position %s", self._ref_start)
+        self._doMoveAbs({self._axis: self._ref_start})
```

Comment:
This should not be done here (as it could well be that it's not yet referenced, so it's not a good idea to move to a "random" position).
It should be done in `doMoveAbs()`, when referencing.

Reply:
Isn't there going to be an issue when we try referencing outside doMoveAbs and the position happens to be negative? Is this dangerous?

Reply:
That's always "dangerous", and it's up to the user to deal with it (when requesting a referencing). That's the case for any actuator anyway. Actually, for the wheel, here, it's not a big deal, as it'll mostly take more time (and can cause an increase of light for a short time). The only thing we can worry about is that when the driver decides to reference automatically, then it shoulnd't do a whole turn.